File: AkkaSingleFs/ReceiveActorWithState.fs
module ReceiveActorWithState

open System
open Akka.Actor

type Messages =
    | Hello of string
    | Swap

type TestActor() as this =
    inherit ReceiveActor()

    let mutable stash: IStash = null

    do this.Normal()

    interface IWithUnboundedStash with
        member this.Stash
            with get () = stash
            and set v = stash <- v

    // First pain point, to use protected member from lambda, it needs a bridge..
    member _.Become x = base.Become(x: Action)

    member this.Normal() =
        this.Receive<Messages> (function
            | Hello s -> printfn $"Hi %s{s}!"
            | Swap -> this.Become(Action(this.Ghost)))

    member this.Ghost() =
        this.Receive<Messages> (function
            // Second pain point, call self implementation of interface is tricky!
            | Hello _ -> (this :> IWithUnboundedStash).Stash.Stash()
            | Swap ->
                this.Become(Action(this.Normal))
                stash.UnstashAll())

let Example (actor_system: ActorSystem) =
    let a = actor_system.ActorOf(Props.Create<TestActor>(), "test-actor")
    printfn "Normal mode"
    a.Tell(Hello "Hello")
    a.Tell(Hello "World")

    printfn "Press ENTER to swap"
    Console.ReadLine() |> ignore
    a.Tell Swap

    printfn "Ghost mode"
    a.Tell(Hello "Keep silent")
    a.Tell(Hello "message after swap")

    printfn "Press ENTER to swap again"
    Console.ReadLine() |> ignore
    a.Tell Swap

    printfn "Press ENTER to end"
    Console.ReadLine() |> ignore
    actor_system.Terminate() |> ignore

File: AkkaSingle/ReceiveActorWithState.cs
using System;
using Akka.Actor;
using AkkaSingle.Messages;

namespace AkkaSingle
{
    namespace Messages
    {
        public sealed record Hello(string Message);
        public sealed record Swap;
    }
    
    public sealed class TestActor : ReceiveActor, IWithUnboundedStash
    {
        public TestActor() {
            Normal();
        }

        public IStash Stash { get; set; } = null!;

        void Normal() {
            Receive<Hello>(h => Console.WriteLine($"Hi {h.Message}!"));
            Receive<Swap>(_ => Become(Ghost));
        }

        void Ghost() {
            Receive<Hello>(_ => Stash.Stash());
            Receive<Swap>(_ => {
                Become(Normal);
                Stash.UnstashAll();
            });
        }
    }

    public static class ReceiveActorWithState
    {
        public static void Example(ActorSystem actorSystem) {
            var a = actorSystem.ActorOf(Props.Create<TestActor>(), "test-actor");
            Console.WriteLine("Normal mode");
            a.Tell(new Hello("Hello"));
            a.Tell(new Hello("World"));
            
            Console.WriteLine("Press ENTER to swap");
            Console.ReadLine();
            a.Tell(new Swap());
            
            Console.WriteLine("Ghost mode");
            a.Tell(new Hello("Keep silent"));
            a.Tell(new Hello("message after swap"));
            
            Console.WriteLine("Press ENTER to swap");
            Console.ReadLine();
            a.Tell(new Swap());
            
            Console.WriteLine("Press ENTER to end");
            Console.ReadLine();
            actorSystem.Terminate();
        }
    }
}